reference/factorization/par_ict_kernels.cpp

```diff
@@ -0,0 +1,209 @@
+#include <unordered_set>
+
+
+#include <ginkgo/core/base/math.hpp>
+#include <ginkgo/core/matrix/coo.hpp>
+#include <ginkgo/core/matrix/csr.hpp>
+
+
+#include "core/base/utils.hpp"
+#include "core/components/prefix_sum.hpp"
+#include "core/matrix/coo_builder.hpp"
+#include "core/matrix/csr_builder.hpp"
+#include "reference/factorization/par_ixt_common.hpp"
+
+
+namespace gko {
+namespace kernels {
+namespace reference {
+/**
+ * @brief The parallel ict factorization namespace.
+ *
+ * @ingroup factor
+ */
+namespace par_ict_factorization {
+
+
+template <typename ValueType, typename IndexType>
+void compute_factor(std::shared_ptr<const DefaultExecutor> exec,
+                    const matrix::Csr<ValueType, IndexType> *a,
+                    matrix::Csr<ValueType, IndexType> *l,
+                    const matrix::Coo<ValueType, IndexType> *)
+{
+    auto num_rows = a->get_size()[0];
+    auto l_row_ptrs = l->get_const_row_ptrs();
+    auto l_col_idxs = l->get_const_col_idxs();
+    auto l_vals = l->get_values();
+    auto a_row_ptrs = a->get_const_row_ptrs();
+    auto a_col_idxs = a->get_const_col_idxs();
+    auto a_vals = a->get_const_values();
+
+    for (size_type row = 0; row < num_rows; ++row) {
+        for (size_type l_nz = l_row_ptrs[row]; l_nz < l_row_ptrs[row + 1];
+             ++l_nz) {
+            auto col = l_col_idxs[l_nz];
+            // find value from A
+            auto a_begin = a_row_ptrs[row];
+            auto a_end = a_row_ptrs[row + 1];
+            auto a_nz_it =
+                std::lower_bound(a_col_idxs + a_begin, a_col_idxs + a_end, col);
+            auto a_nz = std::distance(a_col_idxs, a_nz_it);
+            auto has_a = a_nz < a_end && a_col_idxs[a_nz] == col;
+            auto a_val = has_a ? a_vals[a_nz] : zero<ValueType>();
+            // accumulate l(row,:) * l(col,:) without the last entry l(col, col)
+            ValueType sum{};
+            IndexType lt_nz{};
+            auto l_begin = l_row_ptrs[row];
+            auto l_end = l_row_ptrs[row + 1];
+            auto lt_begin = l_row_ptrs[col];
+            auto lt_end = l_row_ptrs[col + 1];
+            while (l_begin < l_end && lt_begin < lt_end) {
+                auto l_col = l_col_idxs[l_begin];
+                auto lt_row = l_col_idxs[lt_begin];
+                if (l_col == lt_row && l_col < col) {
+                    sum += l_vals[l_begin] * l_vals[lt_begin];
+                }
+                if (lt_row == row) {
+                    lt_nz = lt_begin;
+                }
+                l_begin += (l_col <= lt_row);
+                lt_begin += (lt_row <= l_col);
+            }
+            auto new_val = a_val - sum;
+            if (row == col) {
+                new_val = sqrt(new_val);
+            } else {
+                auto diag = l_vals[l_row_ptrs[col + 1] - 1];
+                new_val = new_val / diag;
+            }
+            if (is_finite(new_val)) {
+                l_vals[l_nz] = new_val;
+            }
```

Comment:
That is difficult to say in general, but I would say that each iteration should improve the approximation, since we can guarantee that the number of "correct" values increases with each iteration (assuming the exact IC decomposition is well-defined)

Reply:
I see, thanks for explanation